src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java

```diff
@@ -0,0 +1,213 @@
+				} else {
+					e.interrupt(Result.NOT_BUILT, new CanceledCause(run));
+				}
+			} else{
+				LOGGER.log(WARNING, "could not cancel an older flow because it has no assigned executor");
+			}
+			LockableResourcesManager.get().removeFromQueue(resourceHolder.required, older);
+		}
+	}
+
+	private static void retry(final String id, final long delay) {
+		// retry only if the the execution of this step has not being somehow stopped
+		StepExecution.applyAll(LockStepExecution.class, new Function<LockStepExecution, Void>() {
+			@Override
+			public Void apply(@Nonnull LockStepExecution execution) {
+				if (execution.id.equals(id)) {
+					execution.tryLock(delay);
+				}
+				return null;
+			}
+		});
+	}
+
+	private static final class Callback extends BodyExecutionCallback.TailCall {
+
+		private final LockableResourcesStruct resourceHolder;
+		private transient Run<?, ?> run;
+		private final String buildExternalizableId;
+
+		Callback(LockableResourcesStruct resourceHolder, Run<?, ?> run) {
+			// It's granted to contain one item (and only one for now)
+			this.resourceHolder = resourceHolder;
+			this.run = run;
+			this.buildExternalizableId = run.getExternalizableId();
+		}
+
+		protected void finished(StepContext context) throws Exception {
+			unlock(context);
+		}
+
+		private void unlock(StepContext context) throws IOException, InterruptedException {
+			if (run == null && buildExternalizableId != null) {
+				run = Run.fromExternalizableId(buildExternalizableId);
+			}
+			LockableResourcesManager.get().unlock(resourceHolder.required, run);
+			// It's granted to contain one (and only one for now)
+			context.get(TaskListener.class).getLogger().println("Lock released on resouce [" + resourceHolder.required.get(0) + "]");
+			LOGGER.finest("Lock released on [" + resourceHolder.required.get(0) + "]");
+		}
+
+		private static final long serialVersionUID = 1L;
+
+	}
+
+	@Override
+	public void stop(Throwable cause) throws Exception {
+		if (body != null) {
+			body.cancel(cause);
+		}
+		if (task != null) {
+			task.cancel(false);
+		}
+		getContext().onFailure(cause);
+	}
+
+	/**
+	 * Records that a build was canceled because it reached a milestone but a
+	 * newer build already passed it, or a newer build
+	 * {@link Milestone#wentAway(Run)} from the last milestone the build passed.
+	 */
+	public static final class CanceledCause extends CauseOfInterruption {
+
+		private static final long serialVersionUID = 1;
+
+		private final String newerBuild;
+		private final String cause;
+
+		CanceledCause(Run<?, ?> newerBuild) {
+			this.newerBuild = newerBuild.getExternalizableId();
+			this.cause = null;
+		}
+
+		CanceledCause(String cause) {
+			this.cause = cause;
+			this.newerBuild = null;
+		}
+
+		public Run<?, ?> getNewerBuild() {
+			return Run.fromExternalizableId(newerBuild);
+		}
+
+		@Override
+		public String getShortDescription() {
+			if (newerBuild != null) {
+				return "Superseded by " + getNewerBuild().getDisplayName();
```

Comment:
It would be nice to know that this was build was superseded because the ${%Limit waiting builds} field was exceeded. Should also include the name of the resource which was locked. Otherwise, I have no idea why a build was canceled?